src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -157,6 +157,12 @@ class ArrayDeque[A] private(var array: Array[AnyRef], var start: Int, var end: I
     }
   }
 
+  override def reverse = {
+    val r = new ArrayDeque[A](initialSize = size)
+    indices.foreach({i => r.appendAssumingCapacity(this(size - i - 1))})
```

Comment:
The JVM can be extremely efficient at _simple_ loops over arrays.

I have seen several cases where a while loop over a byte array for copying is marginally FASTER than System.arraycopy **

Just use a loop, the JVM is very good at optimizing them in cases like this.


** A simple loop over bytes _can_ be compiled down to an assembly REP MOVS or similar, if the right conditions are met.  One instruction for the whole thing.  System.arraycopy is implemented with an intrinsic that (last I checked, a couple years back) has to check a couple preconditions before being so optimal and dispatches slightly differently based on the size of the array.   The precondition for using the single assembly move is that the source and destination do not overlap.   If the JIT can see that the source and destination can never be the same array, it can optimize it further than System.arraycopy _or_ unsafe.  Those conditions are not always transparent to the JIT though.  So by default either write a loop or use System.arraycopy and optimize later by looking at what it compiles into.